File: python/2015/day1.py

from collections import Counter


def part1(lines):
    """
    >>> part1(['(())'])
    0
    >>> part1(['()()'])
    0
    >>> part1(['((('])
    3
    >>> part1(['(()(()('])
    3
    >>> part1(['))((((('])
    3
    >>> part1(['())'])
    -1
    >>> part1(['))('])
    -1
    >>> part1([')))'])
    -3
    >>> part1([')())())'])
    -3
    """
    data = lines[0]
    c = Counter(data)
    return c['('] - c[')']


def part2(lines):
    """
    >>> part2([')'])
    1
    >>> part2(['()())'])
    5
    """
    floor = 0
    for index, b in enumerate(lines[0]):
        if b == '(':
            floor += 1
        elif b == ')':
            floor -= 1
        else:
            raise Exception
        if floor < 0:
            return index + 1

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2015/Day1.kt
package io.github.ocirne.aoc.year2015

import io.github.ocirne.aoc.AocChallenge
import java.lang.RuntimeException

class Day1(val lines: List<String>) : AocChallenge(2015, 1) {

    fun countBrackets(line: String): Int {
        return line.count { it == '(' } - line.count { it == ')' }
    }

    override fun part1(): Int {
        return countBrackets(lines[0])
    }

    fun findBasement(line: String): Int {
        var floor = 0
        line.forEachIndexed { index, bracket ->
            run {
                if (bracket == '(')
                    floor++
                if (bracket == ')')
                    floor--
                if (floor < 0)
                    return index + 1
            }
        }
        throw RuntimeException()
    }

    override fun part2(): Int {
        return findBasement(lines[0])
    }
}
